### PR TITLE
flutter: add aarch64-darwin support

### PR DESCRIPTION
**Pull request by tanshihaj**
###### Description of changes

Add aarch64-darwin support for flutter v3.0.0+ (versions below does not support aarch64-darwin, only x86_64-darwin). All changes are trivial except `./patches/flutter3/move-cache.patch` these hunk:
https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/compilers/flutter/patches/flutter3/move-cache.patch#L125-L140
causes flutter on arm64 macos use some x86_64 tools and I cannot run hello world app on macos platform. Simply replacing `_dartSdkPath(_cache)` with `_dartSdkPath(_fileSystem)` works for me.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tanshihaj** [comment]
It seems that `podhelper.rb` also requires some patches to find correct path with artifacts.

File: pkgs/development/compilers/flutter/default.nix [hunk 1]
```diff
@@ -17,12 +17,21 @@ let
...
     };
-    src = fetchurl {
-      url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_${version}-stable.tar.xz";
-      sha256 = hash;
-    };
+    src = (
+      if stdenv.hostPlatform.isLinux then fetchurl {
+        url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_${version}-stable.tar.xz";
+        sha256 = flutterHash.linux;
+      } else fetchzip {
```
Comment (risicle) [inline pkgs/development/compilers/flutter/default.nix:30]: I don't think it's safe to assume that anything that's not linux is darwin.

**Comment by tanshihaj** [comment]
@justinmoon, great! If you have macos:
- x86_64: Checkout this branch and add to the `pkgs/development/compilers/flutter/default.nix` missing dart and flutter soruces for `x86_64-darwin`. Then try to build some simple hello word project using this flutter.
- aarch64: You do not need any changes on this branch, just try to run hello world project using flutter built from this branch.

**Comment by justinmoon** [comment]
I'm on aarch64. 

>  just try to run hello world project using flutter built from this branch.

Sorry, but how do I do this?

I ran `nix-shell -p flutter -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/pull/196133/head.tar.gz` and building empty project worked on Android emulator, but failed on iOS simulator with [this error](https://gist.github.com/justinmoon/bfd9c15dc7fc8fe78f1f80a440acbe96).

Did I do something wrong?

**Comment by tanshihaj** [comment]
> Sorry, but how do I do this?

I'm neither macOS nor iOS developer, but I've managed to launch hello_world application (the one I've created with `flutter create -t app -d ios fancy_ios_app`) in iOS simulator:
1. Install & config XCode according to https://docs.flutter.dev/get-started/install/macos#install-xcode
2. run nix-shell to get flutter: `nix-shell /Users/me/Documents/code/nixpkgs -A flutter`
3. create hello world project: `flutter create -t app -d ios fancy_ios_app`
4. go to created `fancy_ios_app` folder and create `default.nix`:
```
{ pkgs ? import /Users/me/Documents/code/nixpkgs {} }:
pkgs.stdenvNoCC.mkDerivation {
    name = "fancy_ios_app";
    nativeBuildInputs = [ pkgs.flutter pkgs.cocoapods ];
}
```
5. enter to this environment `nix-shell`
6. launch ios simulator `open -a Simulator` (you may need to config some device)
7. run flutter app`flutter run -d 3D1B33DA-E714-4414-A43C-89A38C929EFE`

The main difference between your environment and my is that you've used nix's build tools and I've used one that supplied with XCode (`pkgs.stdenvNoCC`). Nix does not support iOS platform so it doesn't have `UIKit` framework (and I think not only this one), you have to use compiler and framework supplied with XCode.

**Comment by tanshihaj** [comment]
@risicle Is there any blockers to merge PR?

**Review by ericdallo — APPROVED** [review]

**Comment by tanshihaj** [comment]
I still cannot merge this PR (I think that I don't have enough permissions to do that). Can someone else do it?

**Comment by ericdallo** [comment]
Ping @thiagokokada

**Review by babariviere — APPROVED** [review]

File: pkgs/development/compilers/flutter/default.nix [hunk 2]
```diff
@@ -31,19 +42,33 @@ in
...
   };
 
   v2 = flutterDrv {
     pname = "flutter";
     version = "2.10.5";
     dartVersion = "2.16.2";
-    hash = "sha256-DTZwxlMUYk8NS1SaWUJolXjD+JnRW73Ps5CdRHDGnt0=";
+    flutterHash = {
+      linux = "sha256-DTZwxlMUYk8NS1SaWUJolXjD+JnRW73Ps5CdRHDGnt0=";
+    };
```
Comment (thiagokokada) [inline pkgs/development/compilers/flutter/default.nix:71]:
```suggestion
    flutterHash = {
      x86_64-linux = "sha256-DTZwxlMUYk8NS1SaWUJolXjD+JnRW73Ps5CdRHDGnt0=";
    };
```

File: pkgs/development/compilers/flutter/default.nix [hunk 3]
```diff
@@ -31,19 +42,33 @@ in
     pname = "flutter";
     version = "3.3.3";
     dartVersion = "2.18.2";
-    hash = "sha256-MTZeWQUp4/TcPzYIT6eqIKSPUPvn2Mp/thOQzNgpTXg=";
+    flutterHash = {
+      linux = "sha256-MTZeWQUp4/TcPzYIT6eqIKSPUPvn2Mp/thOQzNgpTXg=";
+      aarch64-darwin = "sha256-apU+jgJgNZBAIffLjh6e6B5szwQsjjve3aVXZDPjD94=";
```
Comment (thiagokokada) [inline pkgs/development/compilers/flutter/default.nix:47]:
```suggestion
    flutterHash = rec {
      x86_64-linux = "sha256-MTZeWQUp4/TcPzYIT6eqIKSPUPvn2Mp/thOQzNgpTXg=";
      aarch64-linux = x86_64-linux;
      aarch64-darwin = "sha256-apU+jgJgNZBAIffLjh6e6B5szwQsjjve3aVXZDPjD94=";
```

File: pkgs/development/compilers/flutter/default.nix [hunk 4]
```diff
@@ -17,12 +17,23 @@ let
...
-      sha256 = hash;
-    };
+    src = rec {
+      x86_64-linux = fetchurl {
+        url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_${version}-stable.tar.xz";
+        sha256 = flutterHash.linux;
+      };
+      aarch64-linux = x86_64-linux;
+      aarch64-darwin = fetchzip {
+        url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/macos/flutter_macos_${if stdenv.hostPlatform.isAarch64 then "arm64_" else ""}${version}-stable.zip";
```
Comment (thiagokokada) [inline pkgs/development/compilers/flutter/default.nix:33]:
```suggestion
        url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/macos/flutter_macos_${lib.optionalString stdenv.hostPlatform.isAarch64 "arm64_"}${version}-stable.zip";
```

File: pkgs/development/compilers/flutter/default.nix [hunk 5]
```diff
@@ -17,12 +17,23 @@ let
...
-    src = fetchurl {
-      url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_${version}-stable.tar.xz";
-      sha256 = hash;
-    };
+    src = rec {
+      x86_64-linux = fetchurl {
+        url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_${version}-stable.tar.xz";
+        sha256 = flutterHash.linux;
+      };
+      aarch64-linux = x86_64-linux;
```
Comment (thiagokokada) [inline pkgs/development/compilers/flutter/default.nix:31]:
```suggestion
      x86_64-linux = fetchurl {
        url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_${version}-stable.tar.xz";
        sha256 = flutterHash.x86_64-linux;
      };
      aarch64-linux = etchurl {
        url = "https://storage.googleapis.com/flutter_infra_release/releases/stable/linux/flutter_linux_${version}-stable.tar.xz";
        sha256 = flutterHash.aarch64-linux;
      };
```

File: pkgs/development/compilers/flutter/default.nix [hunk 6]
```diff
@@ -31,19 +42,33 @@ in
...
-    patches = getPatches ./patches/flutter3;
+    patches = [
+      ./patches/flutter3/disable-auto-update.patch
+      ./patches/flutter3/git-dir.patch
+      (if stdenv.hostPlatform.isLinux then
+        ./patches/flutter3/move-cache-linux.patch
+      else
+        ./patches/flutter3/move-cache-darwin.patch
+      )
+    ];
```
Comment (thiagokokada) [inline pkgs/development/compilers/flutter/default.nix:62]:
```suggestion
    patches = [
      ./patches/flutter3/disable-auto-update.patch
      ./patches/flutter3/git-dir.patch
    ]
    ++ lib.optional stdenv.hostPlatform.isLinux ./patches/flutter3/move-cache-linux.patch
    ++ lib.optional stdenv.hostPlatform.isDarwin ./patches/flutter3/move-cache-darwin.patch;
```

File: pkgs/development/compilers/flutter/flutter.nix [hunk 7]
```diff
@@ -33,12 +33,18 @@
...
 }:
 let
   drvName = "flutter-${version}";
   flutter = stdenv.mkDerivation {
     name = "${drvName}-unwrapped";
 
+    nativeBuildInputs = []
+    ++ lib.optionals stdenv.hostPlatform.isDarwin [
+      darwin.DarwinTools
+    ];
```
Comment (thiagokokada) [inline pkgs/development/compilers/flutter/flutter.nix:46]:
```suggestion
    nativeBuildInputs = lib.optionals stdenv.hostPlatform.isDarwin [
      darwin.DarwinTools
    ];
```

File: pkgs/development/compilers/flutter/flutter.nix [hunk 8]
```diff
@@ -151,12 +157,18 @@ let
 self = (self:
 runCommand drvName
 {
-  startScript = ''
+  startScript = (if stdenv.hostPlatform.isLinux then ''
```
Comment (thiagokokada) [inline pkgs/development/compilers/flutter/flutter.nix:160]:
```suggestion
  startScript = lib.optionalString stdenv.hostPlatform.isLinux ''
```

File: pkgs/development/compilers/flutter/flutter.nix [hunk 9]
```diff
@@ -151,12 +157,18 @@ let
...
     #!${bash}/bin/bash
     export PUB_CACHE=''${PUB_CACHE:-"$HOME/.pub-cache"}
     export ANDROID_EMULATOR_USE_SYSTEM_LIBS=1
     ${fhsEnv}/bin/${drvName}-fhs-env ${flutter}/bin/flutter --no-version-check "$@"
-  '';
+  '' else ''
```
Comment (thiagokokada) [inline pkgs/development/compilers/flutter/flutter.nix:165]:
```suggestion
  ''
  + lib.optionalString stdenv.hostPlatform.isDarwin ''
```

**Review by thiagokokada — CHANGES_REQUESTED** [review]
Looks mostly good, but I think we shouldn't assume `aarch64-linux == x86_64-linux` (this makes the code looks weird).

Also, using `lib.optional` and `lib.optionalString` for the platform specific code is clearer IMO.

File: pkgs/development/compilers/flutter/default.nix (hunk 2 above)
Comment (tanshihaj) [reply pkgs/development/compilers/flutter/default.nix:71]: It is hash for for arch-independent source of flutter. I think we need to leave at as linux.

**Comment by tanshihaj** [comment]
I've changed everything execpt `linux` -> `x86_64-linux` since I think it is misleading to mark arch-independent code as `x86_64` and `aarch64`.

**Comment by thiagokokada** [comment]
It seems the CI is failing thanks to sha256 hash mismatch. Can you double check?

**Comment by tanshihaj** [comment]
Yes, after rebase to latest master I didn't change flutter archive hash for linux platform.

File: pkgs/development/compilers/flutter/default.nix (hunk 2 above)
Comment (thiagokokada) [reply pkgs/development/compilers/flutter/default.nix:68]: We should individually list all arches here them. Having two different formats is really confusing.

Comment (thiagokokada) [reply pkgs/development/compilers/flutter/default.nix:68]: If you don't want to repeat the hashes, you can do it like this:

```suggestion
    flutterHash = rec {
      x86_64-linux = "sha256-DTZwxlMUYk8NS1SaWUJolXjD+JnRW73Ps5CdRHDGnt0=";
      aarch64-linux = x86_64-linux;
    };
```

File: pkgs/development/compilers/flutter/default.nix [hunk 10]
```diff
@@ -31,19 +42,30 @@ in
     pname = "flutter";
     version = "3.3.8";
     dartVersion = "2.18.4";
-    hash = "sha256-QH+10F6a0XYEvBetiAi45Sfy7WTdVZ1i8VOO4JuSI24=";
+    flutterHash = {
+      linux = "sha256-QH+10F6a0XYEvBetiAi45Sfy7WTdVZ1i8VOO4JuSI24=";
```
Comment (thiagokokada) [inline pkgs/development/compilers/flutter/default.nix:46]: Same as here: https://github.com/NixOS/nixpkgs/pull/196133/files#r1013277752

Comment (tanshihaj) [reply pkgs/development/compilers/flutter/default.nix:46]: yes, now I understand where is the issue - I broke linux builds.

Comment (tanshihaj) [reply pkgs/development/compilers/flutter/default.nix:46]: But I cannot understand how. I've tried to build `flutter` & `flutter2` on  cd09fd5 commit on linux virtual machine and it builds ok. Can you show how and where CI fails?

Comment (tanshihaj) [reply pkgs/development/compilers/flutter/default.nix:46]: @thiagokokada ?

**Comment by devholic** [comment]
@tanshihaj @thiagokokada Is there any updates on this PR?

**Comment by thiagokokada** [comment]
> @tanshihaj @thiagokokada Is there any updates on this PR?

Needs a rebase at least.

**Comment by tanshihaj** [comment]
And this question https://github.com/NixOS/nixpkgs/pull/196133#discussion_r1046839652 is still actual.

File: pkgs/development/compilers/flutter/default.nix (hunk 10 above)
Comment (thiagokokada) [reply pkgs/development/compilers/flutter/default.nix:46]: I was not talking about the CI, it seems ok to me at least.

Comment (devholic) [reply pkgs/development/compilers/flutter/default.nix:46]: @tanshihaj @thiagokokada Not sure I understand correctly, but I created PR, which may resolves this discussion for supporting both aarch and x86 based on this PR. Could you take a look? 

https://github.com/NixOS/nixpkgs/pull/210067

**Pull request closed without merging**